### ISSUE TITLE
This condition will always return 'false' since the types have no overlap.

### ISSUE THREAD

**Issue by Busyrev** [microsoft/TypeScript#25642]
**TypeScript Version:**  3.0.0-dev.20180712
Error: This condition will always return 'false' since the types 'A' and 'I' have no overlap.
**TypeScript Version:**  2.9.2
Error:  Operator '===' cannot be applied to types 'A' and 'I'

**Search Terms:** 
Operator '===' cannot be applied to types 
This condition will always return 'false' since the types have no overlap.
**Code**

```ts
class A { 
  private field: string;
}
interface I { 
  pub: string;
}
class B extends A implements I { 
  pub: string;
}
let b = new B();
let a: A = b
let i: I = b;
if (a === i) { // on TS 2.9.2: Operator '===' cannot be applied to types 'A' and 'I'
// on TS 3.0.0-dev.20180712:  This condition will always return 'false' since the types 'A' and 'I' have no overlap.
  console.log('qq');
}
```

**Expected behavior:**
no error
**Actual behavior:**
error
**Playground Link:** [link](https://www.typescriptlang.org/play/index.html#src=class%20A%20%7B%20%0D%0A%20%20private%20field%3A%20string%3B%0D%0A%7D%0D%0Ainterface%20I%20%7B%20%0D%0A%20%20pub%3A%20string%3B%0D%0A%7D%0D%0Aclass%20B%20extends%20A%20implements%20I%20%7B%20%0D%0A%20%20pub%3A%20string%3B%0D%0A%7D%0D%0Alet%20b%20%3D%20new%20B()%3B%0D%0Alet%20a%3A%20A%20%3D%20b%0D%0Alet%20i%3A%20I%20%3D%20b%3B%0D%0Aif%20(a%20%3D%3D%3D%20i)%20%7B%20%2F%2F%20on%20TS%202.9.2%3A%20Operator%20'%3D%3D%3D'%20cannot%20be%20applied%20to%20types%20'A'%20and%20'I'%0D%0A%2F%2F%20on%20TS%203.0.0-dev.20180712%3A%20%20This%20condition%20will%20always%20return%20'false'%20since%20the%20types%20'A'%20and%20'I'%20have%20no%20overlap.%0D%0A%20%20console.log('qq')%3B%0D%0A%7D)

**Related Issues:** no

This case found from real project. Class and interface may be implemented and extended by another class, and in this case this error is pointless.

**Comment by Busyrev** [comment]
Another case for classes without interfaces

```
class A { 
  private field: string;
}
class B { 
  public anotherField: string;
}
class C extends A implements B { 
  public anotherField: string;
}
let c = new C();
let a: A = c
let b: B = c;
if (a === b) { // on TS 2.9.2: Operator '===' cannot be applied to types 'A' and 'I'
// on TS 3.0.0-dev.20180712:  This condition will always return 'false' since the types 'A' and 'I' have no overlap.
  console.log('qq');
}
```

[Playground](https://www.typescriptlang.org/play/#src=class%20A%20%7B%20%0D%0A%20%20private%20field%3A%20string%3B%0D%0A%7D%0D%0Aclass%20B%20%7B%20%0D%0A%20%20public%20anotherField%3A%20string%3B%0D%0A%7D%0D%0Aclass%20C%20extends%20A%20implements%20B%20%7B%20%0D%0A%20%20public%20anotherField%3A%20string%3B%0D%0A%7D%0D%0Alet%20c%20%3D%20new%20C()%3B%0D%0Alet%20a%3A%20A%20%3D%20c%0D%0Alet%20b%3A%20B%20%3D%20c%3B%0D%0Aif%20(a%20%3D%3D%3D%20b)%20%7B%20%2F%2F%20on%20TS%202.9.2%3A%20Operator%20'%3D%3D%3D'%20cannot%20be%20applied%20to%20types%20'A'%20and%20'I'%0D%0A%2F%2F%20on%20TS%203.0.0-dev.20180712%3A%20%20This%20condition%20will%20always%20return%20'false'%20since%20the%20types%20'A'%20and%20'I'%20have%20no%20overlap.%0D%0A%20%20console.log('qq')%3B%0D%0A%7D)

**Comment by mhegazy** [comment]
//cc @RyanCavanaugh and @DanielRosenwasser for error message details.

**Comment by mhegazy** [comment]
The rule for the type assertion is that the compiler can verify that either of the two types is comparable to the other. In simpler terms, can you assign one to the other either upcast or downcast. if neither is true, then it is likely an error.. e.g. `3 === "string"`.

**Comment by AnyhowStep** [comment]
On one hand, it looks like a code smell to compare types like this.

On the other, if such a check passes, I would expect it to behave like some sort of type guard, where both `a` and `i` are now of type `A & I` or something to that effect.

**Comment by Busyrev** [comment]
In simple cases yes, looks good. But in some cases it alerts on absolutely correct code. Found this case in complicated code that works with heirechical system using abstract interfaces. All parent - child links is interfaces, and nodes implemets this interfaces. 
Think thak this error should be emitted if both arguments primitive or nominal types only. If type of one side is not nominal, its incorrect to prevent equality check.

**Comment by vovkasm** [comment]
@RyanCavanaugh And how do you propose to correct these `"incorrect" errors`? What developers should do?

**Comment by Busyrev** [comment]
[Oops!...I Did It Again](https://github.com/Busyrev/TypeScript/blob/publicChecker/src/compiler/program.ts#L1524)

**Comment by AntiMoron** [comment]
![image](https://user-images.githubusercontent.com/6587734/49511646-ac232e80-f8c6-11e8-8f2d-e30e4f157589.png)

How you guys solve this?

**Comment by AnyhowStep** [comment]
@AntiMoron 
Shouldn't it be `process.env.<someIdentifier> == 'development'`?

**Comment by AntiMoron** [comment]
@AnyhowStep  Thank you...I redefined `process.env` as a string...
Seems to be my problem.

**Comment by vikrampariharcodereadr** [comment]
> ![image](https://user-images.githubusercontent.com/6587734/49511646-ac232e80-f8c6-11e8-8f2d-e30e4f157589.png)
> 
> How you guys solve this?

use String(process.env) == 'development'  

Typecasting :)

**Comment by thinwire** [comment]
How about giving us the ability to silence an error - in my case I have a build script that generates code that contains constants based on the target platform (browser, android, ios...), and in this case `if(BUILDTYPE === 'ANDROID')` generates this error when the condition is false.

**Comment by RyanCavanaugh** [comment]
> How about giving us the ability to silence an error 

`// @ts-ignore`

**Comment by thinwire** [comment]
Derp. Seems my google-fu failed me again. Thank you, Ryan.

**Comment by D252874** [comment]
is there a way to silence the error or an easy way to fix the error. I am using geolocation and need to check if the connection was good it just gives a string and it also works but it still gives the error.

**Comment by Drageaux** [comment]
> ![image](https://user-images.githubusercontent.com/6587734/49511646-ac232e80-f8c6-11e8-8f2d-e30e4f157589.png)
> 
> How you guys solve this?

If you're using node, it's process.env.NODE_ENV. It's a string. 



> > ![image](https://user-images.githubusercontent.com/6587734/49511646-ac232e80-f8c6-11e8-8f2d-e30e4f157589.png)
> > How you guys solve this?
> 
> use String(process.env) == 'development'
> 
> Typecasting :)

process.env is an interface ProcessEnv with `[key: string]: string | undefined` property. Why you're typecasting something totally unrelated from a string to be a string is beyond me.

please use `process.env.<someIdentifier>`

**Comment by tenedor** [comment]
Here's another variant of this problem, in case it's useful. I have code where `T` and `U` may be the same type or different types.

``` typescript
const x = <T extends SomeType, U extends SomeType> (t: T, u: U) => {
    return t === u;
}
```

**Comment by Aloso** [comment]
Here's another simplified example:

```ts
import { Bar1, Bar2 } from './foo'

class Baz implements Bar1, Bar2 {
    bar1() {}
    bar2() {}
}

const baz = new Baz()
const bar1: Bar1 = baz
const bar2: Bar2 = baz

if (bar1 === bar2) { // ERROR!
}
```

foo.ts:
```ts
export interface Foo {}

export interface Bar1 extends Foo {
    bar1(): void
}

export interface Bar2 extends Foo {
    bar2(): void
}
```

The comparison `bar1 === bar2` causes the error "This condition will always return 'false' since the types 'Bar1' and 'Bar2' have no overlap."

Surprisingly, this error only appears if the interfaces are in a **different module**.

**Comment by HemalR** [comment]
Pretty new to Typescript and I came across what _might_ be this error. I've got the code below (screenshots underneath). Am I falling for a 'gotcha' somewhere? Is this expected behaviour or am I doing something wrong that I am being blind to?

Thanks in advance for any pointers!

```
// Both triggerVar and userVarId are typed as a string, and TS seems to know this
if (!triggerVar === userVarId) {
  return false;
}
```

![image](https://user-images.githubusercontent.com/15634303/67654290-036f3900-f988-11e9-8886-8ac71b4babc2.png)

![image](https://user-images.githubusercontent.com/15634303/67654325-1eda4400-f988-11e9-9a1c-3f172353fd1b.png)

**Comment by snarfblam** [comment]
What irks me here is that the error message implies that the code is provably incorrect ("This condition **will always** return 'false'"). It is clearly demonstrated that such a comparison may be perfectly valid.

The premise of the error is not factual, it's heuristic, and that should have been reflected in the error message. "This condition **may** always return 'false' since the **declared** types have no overlap."

It might sound like a nitpick, but the phrasing cues the user on how to resolve the issue. It should reduce confusion instead of creating confusion.

**Comment by niezw** [comment]
> ![image](https://user-images.githubusercontent.com/6587734/49511646-ac232e80-f8c6-11e8-8f2d-e30e4f157589.png)
> 
> How you guys solve this?

import _ from 'lodash';
if(_.isEqual(process.env,'development')){
   than .....
}

**Comment by snarfblam** [comment]
@niezw A correct answer was already provided [above](https://github.com/microsoft/TypeScript/issues/25642#issuecomment-502238607). All you're doing is misusing a library function because it happens to circumvent the type system and suppress an error message, when in fact the error message is correctly identifying a problem with the program logic.

`process.env` is not a string. It is an object. If for some strange reason it ever *were* a string, the correct solution would be to fix the type declaration or, as a bandaid, use `//@ts-ignore` or perform a cast: `if(process.env as unknown as string === 'development') {...`. But due consideration should be given to the fact that TS is reporting `process.env` as type `ProcessEnv` and not `string`.

**Comment by prog-24** [comment]
Cast whatever value to any like this

`const a:any = problematicvalue;`

You can then compare as normal

**Comment by snarfblam** [comment]
@prog-24 The problem isn't that we don't know how to work around the error. It's an issue of whether the error is meaningful and instructive. Blindly casting is asking for trouble. In the above case of `process.env`, the error message is correctly identifying a problem. However, a more accurate error message would help us deduce whether there is a flaw in our logic or if it is a false positive that we should work around via a cast to `any` or `unknown`

**Comment by prog-24** [comment]
@snarfblam I think it is one of those compiler things that are really annoying but somehow necessary evil. I had a similar issues that caused me to produce this work around. There are times when yes you can make the necessary changes but if you are sure that it would not cause any issues for instance a simple `string` to `struct` string comparison as was my case then you can safely make the adjustments. 

I  think the error message is clear but unhelpful which I think is what you are pointing out.

**Comment by snarfblam** [comment]
Well, no. I think it is helpful but unclear. As I said above, the error is heuristic (the heuristic being that comparing two values whose declared types have no overlap is moderately indicative of a potentially hard-to-diagnose programmer mistake—very helpful), and it could be more clearly stated along the lines of, "This condition **may** always return 'false' since the **declared** types have no overlap."

Currently the error misstates the problem, making it more difficult to understand the root cause. Typescript errors can be confusing to parse (it's not that unusual to get a single error that is several lines of deeply nested symbol soup). If the error message isn't **clear and accurate**, it can be extremely difficult to work out. I'm imagining a scenario where a programmer develops a knee-jerk habit of casting to `any` whenever this error message pops up because it's phrasing doesn't make sense (or more to the point, it states a falsehood), negating the genuine utility of the error.

**Comment by kevincox** [comment]
This is a very poorly worded "error". For example the following code: `"1" == 1` triggered the error. While the error claims "This condition will always return 'false'" it in fact evaluates to true. This is incredibly misleading.

[Playground Link](https://www.typescriptlang.org/v2/en/play?ssl=1&ssc=1&pln=1&pc=9#code/EQRmAIF5PEg)

**Comment by BertCatsburg** [comment]
I got the same error-message with the following code: 
```
let x = null;
if ( some-condition ) {
  x = 0;
} else {
  x = 1;
}
if ( x === 0 ) {
  // Do something
}
```
On the x===0 condition I got the error, which was strange in my view because I was sure the x is either 0 or 1. 
But the culprit was the initialization to null. 
Changed that to: 
```
let i;
```
And it solved the problem.
If it's a casting problem then its still strange because the program has x casted to an int the moment of the first condition. (x = 0 or x = 1)

**Comment by victor-zy** [comment]
`typeof a === b`,试一下咯

**Comment by douglasmatosdev** [comment]
I solved this situation
![image](https://user-images.githubusercontent.com/57239907/87698572-0bffdd80-c76a-11ea-9987-88d2eba44e29.png)

using `String()`
![image](https://user-images.githubusercontent.com/57239907/87698692-36ea3180-c76a-11ea-9605-9761958d2e94.png)

**Comment by snarfblam** [comment]
> I solved this situation

This is just another example of someone introducing a bug into their program to try to fix this error message. There is generally no good reason to use `String`, and now you have a function that is supposed to return a `boolean` that is instead returning `String('true')` or `String('false')`, both of which are objects, thus **truthy**.

@RyanCavanaugh Is there any chance we can re-open this issue to consider re-wording the error message? I think it should convey:
- That the comparison is *probably* erroneous
- That the comparison may also merely require a cast to help unconfuse the compiler.

Perhaps something along the lines of:

> This condition may always return 'false' since the declared types have no overlap. Consider casting if the compared values are known to be of compatible types.

**Comment by NeliHarbuzava** [comment]
I'm also having this error while doing in-runtime typechecking for strings. I know that by the typing system definition my condition seems to be always true/false. But this is true only for compiling face. In the runtime I get theese values from the user, so I need to check them at a runtime. And in those checking conditions TS says that it's always 'true'/'false' =(

**Comment by Aloso** [comment]
@NeliHarbuzava Your code is wrong, it currently _always_ returns `true`, even if `imageFormat` is `"png"` or `"jpg"`. The correct code would be

```rust
if (typeof imageFormat !== 'string' || (imageFormat !== 'png' && imageFormat !== 'jpg')) {
//                                                            ^^ problem solved
      console.warn('Invalid "imageFormat" property. Expected one of: "png", "jpg".');
}
```

**Comment by snarfblam** [comment]
> `This condition will always return 'true' since the types '"png"' and '"jpg"' have no overlap`

This demonstrates the actual utility of the error message. `imageFormat !== 'png' || imageFormat !== 'jpg'` *will* always return true, because it will always be inequal to at least one of the two ('jpg' or 'png'). It would be nice if TS could distinguish between genuine "will always return true" versus heuristic "may always return true", but I'd settle for any improvement to how the error is reported.

**Comment by OliverEvans96** [comment]
@tenedor - thanks for the succinct  MWE - I'm in the same situation.

> ```ts
> const x = <T extends SomeType, U extends SomeType> (t: T, u: U) => {
>     if (t === u) {
>         ...
>     } else {
>         ...
>     }
> }
> ```

One simple workaround is to set `u: T | U`, although that may have unintended consequences I'm not aware of.

**Comment by galenwinsor** [comment]
As far as I can tell, this is still an issue. As @kevincox points out, the simple expression `'1' == 1` throws this error, which is extremely misleading. @RyanCavanaugh can you clarify why the wording of the error is useful? (I.e., "**will always** return false" as opposed to "**may return** false"). Thank you!

**Comment by RyanCavanaugh** [comment]
See #26592

**Comment by galenwinsor** [comment]
@RyanCavanaugh I guess I don't see how that answers the question. I can see the use of pointing out the potential bug-iness of using `==`, but the _language_ of the error message is misleading and also incorrect, isn't it?

**Comment by galenwinsor** [comment]
Ahhh okay, sorry for the confusion. Thank you!

**Comment by galenwinsor** [comment]
@RyanCavanaugh could you point me to where this check happens in the code?
